_posts/2017-11-30-totw-130.md

```diff
@@ -0,0 +1,185 @@
+
+``` java
+public void f() {
+  com.google.foo.bar.Baz b = new com.google.foo.bar.Baz();
+}
+```
+
+or it is imported, either as a single package member or via wildcard:
+
+``` java
+import com.google.foo.bar.Baz;
+import com.google.foo.bar.*;
+```
+
+In no case is `Baz` looked for outside of the package that is explicitly
+provided: wildcards don't descend into child packages, nor is search extended
+into parent packages. As it turns out, the difference in how parent
+packages/namespaces is handled is fundamental to why structural namespace naming
+(making the namespace structure match the package hierarchy) is a mistake.
+
+## The problem
+
+The fundamental problem for building namespaces out of packages is that we
+rarely rely on fully-qualified lookup in C++, normally writing `std::unique_ptr`
+rather than `::std::unique_ptr`. Coupled with lookup in enclosing namespaces,
+this means that for code in a deeply nested package
+(`::division::section::team::subteam::project`, for example) any symbol that is not
+fully qualified (`std::unique_ptr`) can in fact reference any of
+
+*   `::std::unique_ptr`
+*   `::division::std::unique_ptr`
+*   `::division::section::std::unique_ptr`
+*   `::division::section::team::std::unique_ptr`
+*   `::division::section::team::subteam::std::unique_ptr`
+*   `::division::section::team::subteam::project::std::unique_ptr`
+
+And what's worse: unqualified search starts at the bottom of that list *and
+stops as soon as there is a namespace match*. This means that your build can be
+broken if any of your transitive includes add a previously unused namespace that
+matches the leading namespace of any symbol you use out of an unqualified
+namespace. Strictly speaking, this doesn't even have to be a build break: if
+someone adds something with a matching name and a syntactically-compatible API,
+the implementation of that API may be completely incompatible and cause
+widespread havoc at runtime. Obviously this isn't too bad with `std` - nobody
+should ever be adding a nested namespace `std`, but what about more common
+namespaces? How about things like `testing`?
+
+Names aren't chosen to be unique. Since teams commonly create local utility
+packages to handle common tasks relating to the infrastructure they rely on, we
+wind up with local `util` and `pipeline` packages - and
+sub-namespaces. This is a recipe for unnecessary and unintended collisions.
+
+For comparison, the problem in Java is far reduced: if you wildcard-import from
+two packages in Java and one adds a new symbol with the same name as the other
+package, your build can break. This is easily and completely solved by
+forbidding wildcard imports as is done in many Java styles.
+
+## Two Consistent Options, Three Approaches
+
+There are two features that prevent this build-break-at-a-distance.
+
+*   If no leaf namespace (`search::foo::bar`) matches any top-level namespace
+    (`::bar`) or a sub-namespace of any parent of that leaf (`search::bar`), there
```

Comment:
Nit: "there will be no problems" sounds a bit awkward, but i can live with it.

"no name collisions will occur" (?)